### ISSUE TITLE
Reduce memory footprint of program ast by packing booleans into a bit field

### ISSUE THREAD

**Issue by thebanjomatic** [rollup/rollup#5127]
### Feature Use Case

The ast nodes on the rollup side of things consume a lot of memory for large projects. These are generally speaking already fairly small objects, but in many cases the sheer volume of nodes means that small reductions in object size for commonly appearing nodes can have larger impacts.

For example, on the project I'm currently building, there are 1,309,167 Identifier nodes that are each 128 bytes. Resulting in ~160 MiB of memory for Identifier objects alone (excluding the data they retain through references). Each boolean field on an object currently uses 4 bytes of memory for 1 bit of information.

Looking at the shape of the Identifier node, there are a 3 booleans (included, deoptimized, and isTDZAccess) so we can save 8 bytes (12 - 4) per instance by packing the boolean values into a single 4 byte number field. (Side note: isTDZAccess requires 2-bits of data as it currently is `boolean|null` to allow us to cache the result if it is currently null).

<img width="560" alt="image" src="https://github.com/rollup/rollup/assets/17907922/7fa63886-5a6b-4f31-8368-ef964883f455">

Two of those boolean values exist in the `NodeBase` and `ExpressionEntity` base classes, so this optimization reduces the size of all node classes by 4 bytes at a minimum + 4 bytes for every boolean of their own that gets packed into the field.

Additionally, there are a few other things we can do like only setting esTreeNode if it needs to be retained (most of the time it doesn't), or providing `this.context` through an accessor that that returns `return this._context ?? this.parent.context` that would allow the value to be overridden by node subclasses, but for the default case would save another 4 bytes for each of these two fields.

**I have tested this approach by modifying the rollup js output in node_modules directly, and this appears to have saved about 120mb of AST node data in total according to the heap snapshots.**

After this optimization, Identifier takes 96 bytes per instance vs 128 (32 byte reduction resulting in 40 MiB less heap data)
<img width="551" alt="image" src="https://github.com/rollup/rollup/assets/17907922/9e68d1fd-307e-4012-be77-45a8c827e45f">

This initial investigation only touched `ExpressionEntity`, `NodeBase`, `Identifier`, and `MemberExpression`, but as mentioned previously, `ExpressionEntity` and `NodeBase` are base classes for lots of other node types.

### Feature Proposal

The way I implemented this test was to add a number field `flags` to `ExpressionEntity` along with a `getFlag`/`setFlag` helper function that takes a mask and a value.

Each boolean field/member on the class is then replaced with a getter/setter pair that calls into that helper function using an enum that defines the mask values
```ts
class ExpressionEntity {
    protected flags: number = 0;
    protected getFlag(mask: number): boolean {
        return (this.flags & mask) !== 0
    }
    protected setFlag(mask: number, value: boolean): void {
        const original = this.flags;
        this.flags = value ? (original & ~mask) : (original | mask);
    }
    
    protected get included(): boolean {
        return this._getFlag(FLAG_INCLUDED);
    }
    protected set included(value: boolean): void {
        this._setFlag(FLAG_INCLUDED, value);
    }
    
    // ... more code
}
```

classes that then derive from `ExpressionEntity` then just need to provide those accessors instead of defining fields on the class. Its a little more verbose, but not too bad:

This:
```ts
class NodeBase extends ExpressionEntity {
	protected deoptimized = false;
}
```

becomes this:
```ts
class NodeBase extends ExpressionEntity {
	protected get deoptimized(): boolean {
            return this.getFlag(FLAG_DEOPTIMIZED);
        }
	protected get deoptimized(value: boolean) {
            this.setFlag(FLAG_DEOPTIMIZED, value);
        }

	constructor() {
            // if the previous initializer was !== false then we need to initialize in the constructor,
            // otherwise we can omit this:
            // this.deoptimized = true; 
        }
}
```

In my test project, I only wound up using 8 bits for the mask, but numbers still take up the full 32 bits. I don't anticipate needing more than 32 bits for this process as there aren't that many more booleans, but technically speaking the sub-type bits can overlap between node types if we need to. I'd prefer to avoid that if we can just because it leads to more confusion.

**Comment by thebanjomatic** [comment]
Capturing more stats for comparison. These heap snapshots take a very long time to create, so I just wanted to make sure I have at least screenshots saved before closing out of my dev tools window...

Before:
<img width="425" alt="image" src="https://github.com/rollup/rollup/assets/17907922/24f20502-2d5b-48c3-82ce-e707cf8d3ac8">

After:
<img width="428" alt="image" src="https://github.com/rollup/rollup/assets/17907922/a435c6ab-25c7-4ce2-ab5e-de16462d2a35">

**Comment by thebanjomatic** [comment]
@lukastaegert I'm willing and able to put up a PR for this if you think it is worth doing and doesn't make the code harder to maintain.

I also saw on the SWC PR that you had a task: `Use one 32bit block for all booleans and fixed strings in any AST node to further reduce size`. I wasn't sure if you were referring to packing the AST in the binary form when transferring the AST between rust and node, or if you meant changing rollups node ast format, but if this overlaps or conflicts with any work you are doing in that branch, please let me know.

Thanks!

**Comment by lukastaegert** [comment]
Hello @thebanjomatic, this is a very interesting analysis and indeed comes at a good time.

> I wasn't sure if you were referring to packing the AST in the binary form when transferring the AST between rust and node, or if you meant changing rollups node ast format

It refers to the former, BUT my hope is that eventually, we can stop copying properties from the binary AST to the Rollup nodes and instead pick AST properties directly from the buffer. Your analysis would further support that this would likely save a lot of memory, especially since the AST buffer does not have any complex hidden data structures attached in the JS runtime.
But this would only have an impact on AST properties while you are referring to the mutable state of the analysis.

While eventually, I would love to put all mutable state into an ArrayBuffer as well, this is far future talk at the moment. Meanwhile, your suggestion would bring tangible improvements for everyone right now.

So yes, a PR is welcome. As for readability, I am fine with have the small indirection, especially if there is a comment somewhere at Node that explains how this saves memory.

I would even be fine with not using set/getFlag but directly do the bit operation each time we need it if we only do it via well-named constants that explain what the current bit means. But only if that would improve performance, maybe you can benchmark if this would have an effect as well? Because that would also have a bearing on how we pick AST flags from the buffer in the future.

**Comment by thebanjomatic** [comment]
> I would even be fine with not using set/getFlag but directly do the bit operation each time we need it if we only do it via well-named constants that explain what the current bit means. But only if that would improve performance, maybe you can benchmark if this would have an effect as well? Because that would also have a bearing on how we pick AST flags from the buffer in the future

I did a cursory investigation of this and found that it didn't make much difference at all. I'm pretty sure v8 is smart enough to inline these small functions or something like that, or at least when they have been called enough times.

The biggest contributor to the performance regression I believe came from the refactor to remove the context reference from the NodeBase class:
```js
get context() { return this.parent.context; }`
```

I wasn't thinking at the time I wrote this, but essentially this is doing N function calls to resolve the context where N is the depth between the node and the top-level Program node :(

I was able to still remove the context reference from the ast node and instead moved the context to the scope. We should have a much smaller number of scopes than we have total ast nodes, so the change is net positive.

I was also able to remove the `this.keys` array on each instance by instead doing `keys[this.type]` to grab the cached keys array for the given node type.

With these change, Identifier is now down to 88 bytes from 128 bytes:
<img width="479" alt="image" src="https://github.com/rollup/rollup/assets/17907922/8d5d80cf-59af-4cf5-827b-31472a752e59">

<img width="408" alt="image" src="https://github.com/rollup/rollup/assets/17907922/b0361e06-7bf5-44b2-8ae7-50ba53fface7">

Total heap size is down 10% / 171 MB (1726MB - 1555MB) when measuring up until right before `await renderChunks` gets called. I think there may be some test failures that I need to look into still, but I should be able to get a PR up by Monday.
